### ISSUE TITLE
add (optional) option use aria2 to manager/accelerate download [$10]

### ISSUE THREAD

**Issue by luzfcb** [yyuu/pyenv#534]
if you show me where in the code is located the logic of the download process, I would be happy to send a pull-request

https://aria2.github.io/

another reason I want a download manager is because where I live (northern region of Brazil), sometimes the Internet connection is unstable, so it's impossible to get finished download, because the connection drops for a few seconds. A download manager like [aria2](https://aria2.github.io/) can deal well with connection failure problems

```bash
fabio@luzfcb:~$ pyenv install 2.7.10 -v
/tmp/python-build.20160202070442.19150 ~
Downloading Python-2.7.10.tgz...
-> https://www.python.org/ftp/python/2.7.10/Python-2.7.10.tgz
curl: (56) SSL read: error:00000000:lib(0):func(0):reason(0), errno 104
error: failed to download Python-2.7.10.tgz

BUILD FAILED (Ubuntu 14.04 using python-build 20151222-8-g5c5205e)

Inspect or clean up the working tree at /tmp/python-build.20160202070442.19150
Results logged to /tmp/python-build.20160202070442.19150.log

Last 10 log lines:
/tmp/python-build.20160202070442.19150 ~
curl: (56) SSL read: error:00000000:lib(0):func(0):reason(0), errno 104

```

**Comment by yyuu** [comment]
@luzfcb How about adding `--retry` for `curl`? I couldn't find equivalent for `wget`, though.

As a temporal workaround, you can place downloaded archives inside `$(pyenv root)/cache` to prevent `pyenv install` to download archives every time. `python-build` (the backend of `pyenv install`) will use cached archive if it is available

https://github.com/yyuu/pyenv/blob/425e44d064f934b78c8858b6d3ad597f46a06fa8/plugins/python-build/bin/python-build#L416

**Comment by luzfcb** [comment]
I believe that the equivalent ```curl --retry``` for `wget` is:

```
       -t number
       --tries=number
           Set number of tries to number. Specify 0 or inf for infinite retrying.  The default is to retry 20 times, with the exception of fatal errors
           like "connection refused" or "not found" (404), which are not retried.
```

but aria2 is more better que wget, for people like me who suffer from bad connections.

**Comment by yyuu** [comment]
Hmm. It seems aria2 has a feature to resume after failed download. It'd be beneficial to add support in https://github.com/rbenv/ruby-build

**Comment by luzfcb** [comment]
`wget` and `aria2` has "`--continue`" option

`curl` is slightly different, has "`-C -`". I'm not sure if it has the same behavior as `--continue` from `wget` and `aria2`

**Comment by thedrow** [comment]
Yes please!
@yyuu @luzfcb I'll buy any of you a beer if you implement it both in ruby-build and python-build.

**Comment by yyuu** [comment]
Umm. It seems that `aria2c` doesn't have feature to do HEAD request only.... It is a requirement for http client used from ruby-build/python-build. I confirmed that the behaviour with aria2c 1.19.0.

**Comment by luzfcb** [comment]
why HEAD feature is necessary?

`aria2c --use-head=true`

https://aria2.github.io/manual/en/html/aria2c.html?highlight=head#cmdoption--use-head

**Comment by yyuu** [comment]
python-build uses `HEAD` request to check if the archive is available on mirror. And, as far as I could confirm, `--use-head=true` seems to request `GET` just after `HEAD`.

https://github.com/yyuu/pyenv/blob/c8ab429d7b0990d901494f9bcd95f6baa5cc51af/plugins/python-build/bin/python-build#L397-L403

**Comment by yyuu** [comment]
It seems `aria2c`'s `--dry-run` isn't a dry run and it will request HEAD to given URL.

**Comment by yyuu** [comment]
@luzfcb I prepared a patch for python-build firstly. I'd prepare a patch for `ruby-build` if it works. Please give it a try.

**Comment by thedrow** [comment]
I hope 10$ is enough for a beer :)

**Comment by thedrow** [comment]
Looks like nothing is broken but I'm not sure how to check that it used aria2c.
Maybe we should log that to somewhere? It's not logged even in verbose mode.

**Comment by yyuu** [comment]
Thanks for bounty :)

Right. It will show nothing at least for now. 
I just disabled verbose logging for similarity between curl and wget, but it might be better to add some logs.

**Comment by thedrow** [comment]
A progress bar would be nice.

**Comment by yyuu** [comment]
I removed `-q` from `aria2c` arguments. It should show the progress of download on verbose mode (`pyenv install -v 3.5.1`).

**Comment by luzfcb** [comment]
@yyuu I plan to test your pull-request today or tomorrow . Thank :-)

**Comment by luzfcb** [comment]
@yyuu I tested, and it works almost as I expected:

1. after download start with `-v` option, for whatever reason, it takes about 1 minute to appear the first log with download information. ( I have no idea how to solve this, I suppose pyenv does something with stdout resulting in this delay )

2. By default, only aria2 enable accelerated download if the file is larger than 20MB, so for my connection to download still too slow. I solved this by manually changing the aria2 download configuration.

I modified the function `http_get_aria2c`, and add the parameters `-x 10 -k 1M -s 10`, where:

* `-x` is number of max connections per server (aria2 default is 1)
* `-k` is minimum split size  (aria2 default is 20MB)
* `-s` number of connections to download (aria2 default is 5)


before:

`aria2c -o "${2:--}" ${options} "$1"`

after:

`aria2c -x 10 -k 1M -s 10 -o "${2:--}" ${options} "$1"`


I running with and without my aria2 configurations changes

with ` -x 10 -k 1M -s 10 ` option: 

3:36 minutes to download, compile and install: https://asciinema.org/a/6tg41eorz1pqj0hqr5ax7ounq 

without: 

21:33 minutes to download, compile and install: https://asciinema.org/a/7xex4x5eniei5ugxnn0i7wdu8

**Comment by yyuu** [comment]
Added support for environment variable of `PYTHON_BUILD_ARIA2_OPTS`. Basically I don't want to add environment specific options as the default option in `python-build`. Please try specifying `aria2c` options via the environment variable.

**Comment by yyuu** [comment]
Basically `python-build` is just redirecting stdout/stderr. I suspect that the cause of the 1 minute delay is in `aria2c` itself. Increasing console logging level (`--console-log-level=debug`?) might help investigation.

**Comment by thedrow** [comment]
Can we have PYTHON_BUILD_ARIA2_OPTS set to -x 10 -k 1M by default.
Otherwise it will be just as slow.

**Comment by wadkar** [comment]
When you merge this, are you thinking of adding an optional dependency to aria2 in the brew formula?

On one hand, it would be nice to indicate the user that we can use aria2 for download if required. On the other hand, the code doesn't depend on `aria2` to work. It just checks if its available and runs it. I am not sure what the norm is in such cases.

**Comment by yyuu** [comment]
The optimal default options for `aria2c` could be diverse per users' environment. For now, I'd like to just keep using default of `aria2c`. It means, the users need to be set their own `PYTHON_BUILD_ARIA2_OPTS` in their environment if needed.

For brew's formula, I may add `aria2` as an optional dependency later. Anyone can also submit PR to them, though 😛

**Comment by yyuu** [comment]
I've merged #539 as the solution for this. Please give it a try.
